### PR TITLE
 Replace ddpg/critic_network.py mlp_layers with encoding networks.

### PR DESCRIPTION
**Pull request by rk37**
do kbanoop's TODO

### CODE REVIEW

File: tf_agents/agents/ddpg/critic_network.py [hunk 1]
```diff
@@ -28,6 +29,9 @@ class CriticNetwork(network.Network):
 
   def __init__(self,
                input_tensor_spec,
+               preprocessing_layers=None,
+               preprocessing_combiner=None,
+               batch_squash=True,
```
Comment (ebrevdo) [inline tf_agents/agents/ddpg/critic_network.py:34]: why do we need this argument publicly available?

**Review by ebrevdo — CHANGES_REQUESTED** [review]
needs unit tests.

File: tf_agents/agents/ddpg/critic_network.py (hunk 1 above)
Comment (rk37) [reply tf_agents/agents/ddpg/critic_network.py:34]: This arguments will pass to encoding_network.EncodingNetwork create an EncodingNetwork  for combi observations and using custom observation preprocessing layers such as Attention.

Comment (ebrevdo) [reply tf_agents/agents/ddpg/critic_network.py:34]: do you ever need to pass a value other than `batch_squash=True` to EncodingNetwork?

Comment (tfboyd) [reply tf_agents/agents/ddpg/critic_network.py:34]: @rk37  Checking in to see if you are still moving forward?  Total cool if you are not, I try to keep things moving and clearly I failed a little here as it has been over a month.

Comment (rk37) [reply tf_agents/agents/ddpg/critic_network.py:34]: > @rk37  Checking in to see if you are still moving forward?  Total cool if you are not, I try to keep things moving and clearly I failed a little here as it has been over a month. 

@tfboyd Sorry, I used the modified code to run my gym env, but I am not sure of its effectiveness in other envs.  I am using pytorch (stable baselines 3) to solve my problem recently. And I will do more tests and add unit test code later.

Comment (tfboyd) [reply tf_agents/agents/ddpg/critic_network.py:34]: @rk37 no problem

**Comment by momja** [comment]
Are there any updates on this? I tried using this for multi-observation SAC, but I've been getting errors when computing the critic loss.

**Comment by tfboyd** [comment]
@momja  What error are you getting? It is nice to have a clean repro and it helps with motivation to get this stuff fixed ASAP. It also allows me to verify if it is fixed by the change or needs another change. If this commit resolves the error, maybe we can resurrect it ASAP. It was really close and then died.

**Comment by ebrevdo** [comment]
+1 would love to see the error.  additionally we have a simplier approach to building Networks now that is not monolithic, using `tf_agents.networks.{Sequential, NestMap, NestFlatten}`.  You can find an example [here](https://github.com/tensorflow/agents/blob/89d189a8e75713b2549fda9e2cc8408e70758265/tf_agents/networks/nest_map_test.py#L72).  We plan to start using this code in more of our examples, but for now it may simplify your model building.

**Comment by ebrevdo** [comment]
We just pushed [an example](https://github.com/tensorflow/agents/commit/b8be56b8a4cfd4085ac68d898389f8c59d4ae804#diff-c078e197628924bc0c7e40ed93dc37c4ae032397e9fff44bda2ee973b83d4097R129) for how to use SAC with Sequential.  We hope to update the other examples, and to do similar with DDPG soon.

**Comment by momja** [comment]
I will be able to report on the error in the next few days, but right now I am testing out this new approach shared by @ebrevdo. I'm receiving some errors while trying to implement my own.

With the modified function:
```
def create_sequential_critic_net():
  value_layer = nest_map.NestMap(
      {
          "patch": patch_pre_layer,
          "color": color_pre_layer,
          "motion": motion_pre_layer
      }
  )
  value_layer = sequential.Sequential([
      value_layer,
      nest_map.NestFlatten(),
      tf.keras.layers.Concatenate(axis=-1)])

  action_layer = tf.keras.layers.Dense(10)

  def sum_value_and_action_out(value_and_action_out):
    value_out, action_out = value_and_action_out
    return tf.reshape(value_out + action_out, [-1])

  return sequential.Sequential([
      nest_map.NestMap((value_layer, action_layer)),
      tf.keras.layers.Lambda(sum_value_and_action_out)
  ])
```

Now if I want to  create variables: `create_sequential_critic_net().create_variables((observation_spec, action_spec))` I get the following error:
```
ValueError: `inputs` and `self.nested_layers` do not have matching structures:
  (DictWrapper({'color': ., 'motion': ., 'patch': .}), .)
vs.
  (., .)
Values:
  (DictWrapper({
'color': <tf.Tensor: shape=(1, 3), dtype=float32, numpy=array([[0.9768766 , 0.04171145, 0.3783009 ]], dtype=float32)>, 
'motion': <tf.Tensor: shape=(1, 3), dtype=float32, numpy=array([[ 5.0205655 , 11.804611  ,  0.68829167]], dtype=float32)>,
'patch': <tf.Tensor: shape=(1, 1, 1, 3), dtype=float32, numpy=array([[[[0.63228047, 0.84821343, 0.74741304]]]], dtype=float32)>}),
<tf.Tensor: shape=(1, 7), dtype=float32, numpy=
array([[-0.0612357 ,  0.27595967,  0.17574099, -0.2420119 ,  2.616015  ,
        -2.1489165 ,  0.69741297]], dtype=float32)>)
vs.
  (<tf_agents.networks.sequential.Sequential object at 0x7fbff42258e0>, <tf_agents.networks.sequential.Sequential object at 0x7fbff4230100>).
```
These structures look like they match to me. This is my current roadblock.

**Comment by ebrevdo** [comment]
I think you're close!  I think your `value_layer` must instead be a dict passed to `NestMap`.  This is the easiest way to do it, and I'm going to be rethinking your use case to see if we can clean up `NestMap` so your version works properly the way you wrote it!

```python
value_layer_dict = {
          "patch": patch_pre_layer,
          "color": color_pre_layer,
          "motion": motion_pre_layer
      }

action_layer = tf.keras.layers.Dense(10)

def sum_value_and_action_out(value_and_action_out):
  value_out_dict, action_out = value_and_action_out
  value_out = tf.concatenate(tf.nest.flatten(value_out_dict), axis=-1)
  return tf.reshape(value_out + action_out, [-1])

  return sequential.Sequential([
      nest_map.NestMap((value_layer_dict, action_layer)),
      tf.keras.layers.Lambda(sum_value_and_action_out)
  ])
```

**Comment by momja** [comment]
> 
> 
> I think you're close! I think your `value_layer` must instead be a dict passed to `NestMap`. This is the easiest way to do it, and I'm going to be rethinking your use case to see if we can clean up `NestMap` so your version works properly the way you wrote it!

You're right! this resolved that problem. I think in the future it would make sense to allow for this kind of structure. I'm still not able to test this out, as I'm still getting an error from `compute_variables` in the SACAgent.
```
AssertionError: Could not compute output KerasTensor(type_spec=TensorSpec(shape=(1, 4096), dtype=tf.float32, name=None), name='flatten/Reshape:0', description="created by layer 'flatten'")
  In call to configurable 'SacAgent' (<class 'tf_agents.agents.sac.sac_agent.SacAgent'>)
```

This I traced back to `patch_pre_layer`, which I provide in the value_out_dict. 

```
patch_pre_layer = tf.keras.models.Sequential(
    [
        tf.keras.layers.Conv2D(8, 3),
        tf.keras.layers.MaxPool2D(),
        tf.keras.layers.Flatten(),
    ]
)
```

However, if I use

```
patch_pre_layer = tf.keras.layers.Flatten()
```
I receive no errors. I am not sure if this is a problem with how SACAgent handles the network, or a problem with my construction.

**Comment by ebrevdo** [comment]
Very interesting - I've never seen that before :(

You could try using `InnerReshape`, which we designed to work for these cases; does [this example](https://github.com/tensorflow/agents/blob/89d189a8e75713b2549fda9e2cc8408e70758265/tf_agents/networks/sequential_test.py#L107) help?.

**Comment by momja** [comment]
I will try that! Have you tested this with different training batch sizes? When I pass an `experience_dataset_fn` to my learner with `batch_size > 1`, I receive the following error

```
ValueError: Input 0 of layer dense_3 is incompatible with the layer: expected axis -1 of input shape to have value 3681 but received input with shape (1, 942336)
```

I would guess this would have something to do with an unspecified batch_size when calling `create_variables` in SACAgent.

**Comment by ebrevdo** [comment]
Is this when using keras Flatten?  keras will flatten inputs of type
[batch, time, ...] and inputs of type [batch, ...] into the same vector.
With tf-agents train() you expect [batch, time, ...] and at inference or in
policy you sometimes expect [batch, ...].  InnerReshape corrects that - it
allows both types of input..

**Comment by tfboyd** [comment]
I have to give up. :-( We made great progress in 2020 as far as accepting PRs and doing so faster, but this didn't make it.

**Pull request closed without merging**